testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java

```diff
@@ -552,6 +553,71 @@ public void testJoinBaseTableWithView()
         assertAccessAllowed(viewOwnerSession, "DROP VIEW " + viewName);
     }
 
+    @Test
+    public void testFunctionOwners()
+    {
+        reset();
+
+        String functionOwner = "function_owner";
+        TrinoPrincipal functionOwnerPrincipal = new TrinoPrincipal(USER, functionOwner);
+
+        systemSecurityMetadata.grantRoles(getSession(), ImmutableSet.of("function_owner_role"), ImmutableSet.of(functionOwnerPrincipal), false, Optional.empty());
+        systemSecurityMetadata.setFunctionOwner(
+                getSession(),
+                new CatalogSchemaFunctionName("memory", "default", "my_test_function_outer"),
+                functionOwnerPrincipal);
+
+        Session functionOwnerSession = TestingSession.testSessionBuilder()
+                .setIdentity(Identity.forUser(functionOwner)
+                        .withEnabledRoles(Set.of("function_owner_role"))
+                        .build())
+                .setCatalog(getSession().getCatalog())
+                .setSchema(getSession().getSchema())
+                .build();
+        assertAccessAllowed(
+                functionOwnerSession,
+                "CREATE FUNCTION memory.default.my_test_function_inner (x integer) RETURNS bigint RETURN x + 42");
+        assertAccessAllowed(
+                functionOwnerSession,
+                "SELECT memory.default.my_test_function_inner(2)");
+
+        assertAccessAllowed("SELECT memory.default.my_test_function_inner(2)");
+
+        assertAccessAllowed(
+                functionOwnerSession,
+                "CREATE FUNCTION memory.default.my_test_function_outer (x integer) RETURNS bigint RETURN x + memory.default.my_test_function_inner(58)");
+
+        assertAccessAllowed(
+                functionOwnerSession,
+                "SELECT memory.default.my_test_function_outer(2)");
+        assertAccessAllowed("SELECT memory.default.my_test_function_outer(2)");
+        assertAccessAllowed(
+                "SELECT memory.default.my_test_function_outer(2)",
+                privilege(getSession().getUser(), "memory.default.my_test_function_inner", EXECUTE_FUNCTION));
+
+        getQueryRunner().getAccessControl()
+                .denyIdentityFunction((identity, function) -> !(identity.getEnabledRoles().contains("function_owner_role_without_access") && "default.my_test_function_inner".equals(function)));
+        systemSecurityMetadata.grantRoles(getSession(), ImmutableSet.of("function_owner_role_without_access"), ImmutableSet.of(functionOwnerPrincipal), false, Optional.empty());
+
+        assertAccessDenied(
+                functionOwnerSession,
+                "SELECT memory.default.my_test_function_outer(2)",
+                "Cannot execute function memory.default.my_test_function_inner");
+        assertAccessDenied(
```

Comment:
It's a fairly complicated test case so maybe it's worth adding a few comments.
As I understand here the default session gets access denied because the function owner got access denied on the inner function, correct?